Replace `DownloadTrackingComposite` with instantiable `DownloadTracker` components

peppy opened the pull request.

peppy committed:
Implement new version of download tracker

peppy committed:
Refactor `ModelDownloader` to allow for different `OnlineID matching`

peppy committed:
Remove old versions of `DownloadTrackingComposite`

peppy committed:
Update `OnlinePlayBeatmapAvailabilityTracker` in line with new download tracker

The structure here has changed a bit - rather than deriving from the
donwload tracker I moved the additional logic into the callbacks. I
think this feels better?

peppy committed:
Remove unnecessary verification flow from score download tracking

peppy committed:
Update remaining usages of download tracking

peppy committed:
Fix some oversights and test failures

peppy committed:
Change test to allow for potentially single-frame delayed events firing

peppy committed:
Fix incorrect initial state

bdach reviewed (commented):
Have read the diff without running. Definitely like the changes, just have some minor points to raise.

I'm not turned off by the "duplication" in beatmap/score download trackers in the slightest, because I feel the composition that this approach enables is a big enough win already. And sometimes there is such a thing as reusing too much and causing everything to become too rigid, which is what the old component was arguably doing.

peppy committed:
Merge branch 'master' into beatmap-refactor/download-tracker

peppy committed:
Simplify `Manager` null check to a single pre-check

peppy committed:
Fix variable name mismatch

peppy committed:
Simplify null check in `ScoreDownloadTracker` and remove unnecessary ID check

peppy committed:
Add inline comment explaining usage of `RequiresChildrenUpdate` override

peppy committed:
Seal implementation and add comment about `LoadComplete` execution order

peppy committed:
Revert to previous exposure of `RoomSubScreen.BeatmapAvailability`

bdach reviewed (dismissed):
I like this. Going to request a once-over from @smoogipoo just to ensure we're on the same page.

peppy committed:
Merge branch 'master' into beatmap-refactor/download-tracker

smoogipoo commented on osu.Game/Online/Rooms/OnlinePlayBeatmapAvailabilityTracker.cs:
```diff
@@ -40,71 +50,67 @@ protected override void LoadComplete()
-
-            var beatmap = Manager.QueryBeatmap(b => b.OnlineBeatmapID == onlineId && b.MD5Hash == checksum);
-            return beatmap?.BeatmapSet.DeletePending == false;
+                AddInternal(downloadTracker);
```
Is this correct...? The `AddInternal()` being inside the progress bindable's callback? Care to explain how this works?

peppy replied:
I think the diff might be confusing you here, github being a bit weird it'd seem.

smoogipoo replied:
Oh, I see. There's two nested `BindValueChange()`s...

smoogipoo commented on osu.Game/Overlays/BeatmapSet/BeatmapSetHeaderContent.cs:
```diff
@@ -222,13 +228,13 @@ private void load(OverlayColourProvider colourProvider)
                 Picker.BeatmapSet = rulesetSelector.BeatmapSet = author.BeatmapSet = beatmapAvailability.BeatmapSet = Details.BeatmapSet = setInfo.NewValue;
                 cover.OnlineInfo = setInfo.NewValue?.OnlineInfo;
 
+                downloadTracker?.RemoveAndDisposeImmediately();
```
Is it even added to the hierarchy in this class at all? Doesn't look to be from what I'm reading.

peppy replied:
Nice catch, this is an oversight.

peppy committed:
Add missing `AddInternal` call for `BeatmapSetHeaderContent`'s download tracker

smoogipoo commented on osu.Game/Screens/OnlinePlay/DrawableRoomPlaylistItem.cs:
```diff
@@ -282,14 +282,15 @@ protected override bool OnClick(ClickEvent e)
             private BeatmapManager beatmapManager { get; set; }
 
-            public override bool IsPresent => base.IsPresent || Scheduler.HasPendingTasks;
+            // required for download tracking, as this button hides itself. can probably be removed with a bit of consideration.
+            public override bool IsPresent => true;
```
https://drive.google.com/file/d/1m7aAbV0Q0xUpKud9mHAuMsiSehbJwb4w/view?usp=sharing

The button is clickable even when it's hidden, making this feel pretty bad... I think this needs to be fixed before merge.

peppy replied:
Any suggestions on how to handle this better? I'd like to avoid having that there in the first place if possible.. Maybe moving the download tracker up one level?

peppy replied:
Actually the fading can probably be moved to the `shakeContainer`, will try that first.

peppy replied:
Ended up fixing this by adding a transition on the width change. Looks pretty cool while also fixing the issue with minimal changes.

peppy committed:
Fix playlist item download button being clickable when it shouldn't

peppy committed:
Fix hash validity check function returning incorrect results after soft deleting a beatmap

The soft-deleted one would potentially match before the newly imported
copy, resulting in a false-positive hash mismatch.
